c2corg_ui/views/document.py

```diff
@@ -413,20 +421,43 @@ def _get_api_cache_key_from_etag(self, etag):
     def _get_response(self, page_html):
         return get_response(self.request, page_html)
 
+    def _redirect(self, id, lang, slug=None):
+        location = ''
+        if slug is None:
+            location = self.request.route_url(
+                self._API_ROUTE + '_view_id_lang', id=id, lang=lang)
+        else:
+            location = self.request.route_url(
+                self._API_ROUTE + '_view', id=id, lang=lang, slug=slug)
+        raise HTTPMovedPermanently(location=location)
```

Comment:
The main problem with using `HTTPMovedPermanently (301)` is that the redirect is actually not permanent. Two examples:

When requesting `/waypoints/1234` (without lang), the redirect depends on the current interface language. Let's say the interface language is `en`, so it will redirect to `/waypoints/1234/en/...`. If the interface language is now changed to `fr`, the browser will directly redirect to `/waypoints/1234/en/...`.

Second problem: If `redirects_to` is set for a document, a permanent redirect to the linked document will be made. But `redirects_to` can be changed! So even if you change `redirects_to`, browsers will still redirect to the old linked document, because they think the redirect is permanent (while it is only temporary!).